### ISSUE TITLE
using tf.function on a model call will not utilize gpu nor VRAM but log placement to gpu

### ISSUE THREAD

**Issue by Apprisco** [tensorflow/tensorflow#56878]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

source

### Tensorflow Version

2.9.1

### Custom Code

Yes

### OS Platform and Distribution

Windows 11

### Python version

3.9.12

### CUDA/cuDNN version

11.2/8.6

### GPU model and memory

3090 24GB

### Current Behaviour?

```shell
Utilize tf.function on a function that calls model(a,b) (aka model.call)
Although eager tensors have all been allocated to the gpu, actual training doesnt happen on the gpu, rather the cpu. Notice 20x slower training speeds than when utilizing nested tf.functions in model.call, which is not allowed in MirroredStrategy.
```

### Standalone code to reproduce the issue

```shell
Model.call:

def call(self,x,y):
    vals=self.disc_gradients(x,y)
    self.gen_gradients(x,y)
    return vals
```
model call wrapper:
```
@tf.function#this is the problematic tf.function, if this tf.function is moved to the below tf.functions it works fine.
def run_batch(self,batch_x,batch_y):
    val=self.model(batch_x,batch_y)
    return val
```
gradients functions:
```
#@tf.function
def get_disc_gradients(self,x,y):
    #calc val
    self.disc_optimizer.apply_gradients(disc_gradients,vars)
    return disc_gradients
```
Obviously this code currently does nothing, it's a minimal reproducible example.
```


### Relevant log output

```shell
WARNING:tensorflow:Using MirroredStrategy eagerly has significant overhead currently. We will be working on improving this in the future, but for now please wrap `call_for_each_replica` or `experimental_run` or `run` inside a tf.function to get the best performance.
```
The above log is relevant as it tries to make us wrap the entire model.call in a tf.function, which is where I discovered this entire issue. This issue is replicateable w/o using MirroredStrategy: all that matters is that tf.function is wrapping a model call.
No other relevant log output could be found.
Only way to notice the issues are the extremely long batch times, along with
the GPU only utilizing 3GB of VRAM compared to a full 20gb when the tf.function is moved.
</details>

**Comment by Apprisco** [comment]
To be more specific, wrapping a training loop for a nested keras.model in any way will cause this issue. Wrapping each individual gradients functions in @tf.function prevents this issue, but is bad since we can't use mirroredstrategy.

**Comment by Apprisco** [comment]
It appears @tf.function placement on nested models matters greatly for gradient calculation: the lack of GPU vram usage is from missing gradients- although it doesn't explain the 700s batch times and clear cpu usage over GPU. Something is seriously wrong here, haha.

**Comment by tilakrayal** [comment]
@Apprisco,
I tried to execute the provided code. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/de0d2a91c5cb9c647a2793b0526751b1/untitled458.ipynb).  The code provided is not complete hence it would be difficult for us to pinpoint the issue. 
Could you share complete stand alone code to replicate the issue or a colab gist with the error reported. Thank you!

**Comment by Apprisco** [comment]
It wouldn't be ideal to share my current code as it is for an upcoming research paper, it may take a bit but let me produce a minimal reproducible example.

**Comment by Apprisco** [comment]
Update: turns out my nested models were not models but layers: this is still an issue regarding tf.function, but different scope.

**Comment by Apprisco** [comment]
Issue went away when we changed to actually using nested models, not gigantic custom layers. I still believe this be an issue but this is no longer one that concerns us. Especially since we moved on to horovod, not mirroredstrategy.

**Comment by Apprisco** [comment]
Error has returned, but again when using nested layers not models haha
